### PR TITLE
Added method to remove meshes from the MeshManager

### PR DESCRIPTION
**Pull request by ahcorde**
Signed-off-by: ahcorde <ahcorde@gmail.com>
# 🎉 New feature

Closes https://github.com/ignitionrobotics/ign-common/issues/37

## Summary

This PR adds and API to remove meshes from the MeshManager. This PR is also related with memory leak https://github.com/ignitionrobotics/ign-gui/issues/208

For now it's a draft PR this may require some more changes.

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge**

### CODE REVIEW

File: graphics/src/MeshManager.cc [hunk 1]
```diff
@@ -241,6 +241,39 @@ const Mesh *MeshManager::MeshByName(const std::string &_name) const
...
+  {
+    delete m.second;
+  }
+  this->dataPtr->meshes.clear();
+}
+
+//////////////////////////////////////////////////
+void MeshManager::RemoveMesh(const std::string &_name)
+{
+  std::cerr << "Trying to remove " << _name << '\n';
```
Comment (mjcarroll) [inline graphics/src/MeshManager.cc:256]: Remove debug print

File: graphics/src/MeshManager.cc [hunk 2]
```diff
@@ -241,6 +241,39 @@ const Mesh *MeshManager::MeshByName(const std::string &_name) const
...
+  bool founded = false;
+  std::string entityToRemove;
+  for (auto m : this->dataPtr->meshes)
+  {
+    std::size_t found = m.second->Name().find(_name);
+    if (found != std::string::npos)
+    {
+      std::cerr << "Mesh founded" << '\n';
```
Comment (mjcarroll) [inline graphics/src/MeshManager.cc:264]: debug print

File: graphics/src/MeshManager.cc [hunk 3]
```diff
@@ -241,6 +241,39 @@ const Mesh *MeshManager::MeshByName(const std::string &_name) const
...
+      founded = true;
+      entityToRemove = m.first;
+    }
+  }
+  if (founded)
+  {
+    std::cerr << "Calling Mesh destructor" << '\n';
```
Comment (mjcarroll) [inline graphics/src/MeshManager.cc:271]: debug print

File: graphics/src/MeshManager.cc [hunk 4]
```diff
@@ -241,6 +241,39 @@ const Mesh *MeshManager::MeshByName(const std::string &_name) const
...
+    delete m.second;
+  }
+  this->dataPtr->meshes.clear();
+}
+
+//////////////////////////////////////////////////
+void MeshManager::RemoveMesh(const std::string &_name)
+{
+  std::cerr << "Trying to remove " << _name << '\n';
+  bool founded = false;
```
Comment (mjcarroll) [inline graphics/src/MeshManager.cc:257]: Two variables `found` vs `founded` is a bit confusing.

File: graphics/include/ignition/common/MeshManager.hh [hunk 5]
```diff
@@ -100,6 +100,9 @@ namespace ignition
       /// \param[in] the mesh to add.
       public: void AddMesh(Mesh *_mesh);
 
+      public: void RemoveMesh(const std::string &_name);
```
Comment (mjcarroll) [inline graphics/include/ignition/common/MeshManager.hh:103]: Add docs

**Review by mjcarroll — CHANGES_REQUESTED** [review]
Looks good overall, I believe that the ABI checker is a false alarm?

**Comment by ahcorde** [comment]
@mjcarroll 

I have been thinking that maybe we should not allow to remove some special meshes. Such the initial ones:

 - axis_head
 - axis_shaft
 - body_cg
 - joint_anchor
 - selection_tube
 - unit_box
 - unit_camera
 - unit_cone
 - unit_cylinder
 - unit_plane
 - unit_sphere

Where do you think is the best place to store this info ?

**Comment by ahcorde** [comment]
@mjcarroll, I have more questions regarding the naming of the meshes, In general there are two types of names:

 - The predefined object and the inserted object have a short name (for example: `unit_box`, `unit_capsule`, etc.
 - When we insert a model with a mesh the name of the mesh is the full path to the mesh in your system. (for example: $HOME/.ignition/fuel/fuel.ignitionrobotics.org/googleresearch/models/threshold_dinner_plate_square_rim_white_porcelain/1/meshes/model.obj

The first kind of objects are very easy to remove but I'm not sure I'm doing the right thing with the second type. For example:

I have a mesh on ignition called `box`, this logic should return `box` or `unit_box` or any other mesh with the keyword `box`in the naming. And this is not what we are looking for. One thing that I can do to improve this is to tokenize the name of the mesh by `/`. But I'm not pretty sure that all meshes will respect `<owner>/models/<name of mesh>/<version>/meshes/<name of file>`. In particular the meshes folder.

To complicate a little bit more the issue, I'm removing this from ign-gazebo using the visual's name, but probably this name will not match with name of the mesh in many cases.

**Comment by chapulina** [comment]
@ahcorde  / @mjcarroll , what's the status of this PR?

I'm not sure if this has been brought up, but we have to be careful when removing meshes because they may be reused by many models.

**Review by chapulina — APPROVED** [review]
LGTM. We need to be careful downstream when using these functions though. Meshes are often reused and removing one of them may have unintended consequences.

**Comment by osrf-triage** [comment]
This pull request has been mentioned on **Gazebo Community**. There might be relevant details there:

https://community.gazebosim.org/t/new-ignition-releases-2022-01-24-citadel-edifice-fortress/1241/1

**Pull request merged**